### ISSUE TITLE
Incorrect output when generating more than 64 bytes

### ISSUE THREAD

**Issue by camerondm9** [mheyman/Isopoh.Cryptography.Argon2#47]
When comparing outputs and performance between this library and [Konscious.Security.Cryptography.Argon2](https://www.nuget.org/packages/Konscious.Security.Cryptography.Argon2), I noticed that your output is different when requesting more than 64 bytes. I verified that your output is also different from that produced by [antelle.net/argon2-browser](https://antelle.net/argon2-browser/), which means there is probably a bug in this library. This difference is present for all addressing modes (Argon2i, Argon2id, Argon2d).

There also isn't any kind of error when setting the memory cost to less than 8 (the minimum in the spec), but your output is different from Konscious.Security.Cryptography.Argon2 again. Not sure whose bug it is, but if you want to match the spec you should probably throw an ArgumentException when memory cost < 8.

**Comment by mheyman** [comment]
Thanks! Will look into it...

**Comment by mheyman** [comment]
I've looked into it...

- Memory cost [translates into a target count of memory blocks](https://mheyman.github.io/Isopoh.Cryptography.Argon2/api/Isopoh.Cryptography.Argon2.Argon2Config.html#Isopoh_Cryptography_Argon2_Argon2Config_MemoryCost) to use for hashing. The actual memory block count minimum is `2*`[`SyncPointCount`](https://mheyman.github.io/Isopoh.Cryptography.Argon2/api/Isopoh.Cryptography.Argon2.Argon2.html#Isopoh_Cryptography_Argon2_Argon2_SyncPointCount)`*`[`Lanes`](https://mheyman.github.io/Isopoh.Cryptography.Argon2/api/Isopoh.Cryptography.Argon2.Argon2Config.html#Isopoh_Cryptography_Argon2_Argon2Config_Lanes) (SyncPointCount is defined as `4` in the spec which is where the `8` comes from when `Lanes=1`). The memory blocks actually used gets clipped to a valid range no matter the target memory cost. Oh, and a memory block is 1024 bytes _and_ the _memory size_ mentioned in the spec is only a target memory size too, it will get floored to `4*Lanes` (and _memory size_ is in _blocks_, not bytes).
   -  I originally wrote this code a long time ago based on an early implementation written in C that, I think, became the first reference implementation. I remember getting confused by the "memory cost" verbiage back then so maybe I should have picked a different name but I fear the ship has sailed on that... I've checked the spec and it only mentions "memory cost" as a security consideration (and that is only in more recent versions of the spec).
   - &lt;rant&gt;I feel like if a coder would have been involved in Argon2 from the beginning, the configuration parameters wouldn't be so messy&lt;/rant&gt;.
- I pulled in the [512-byte tag](https://github.com/kmaragon/Konscious.Security.Cryptography/blob/master/Konscious.Security.Cryptography.Argon2.Test/Argon2Tests.cs) test vectors from Konscious.Security.Cryptography.Argon2 and the Isopoh implementation passes that.
   -  I'm still looking into whether calling in a different way to get more than 64-bytes of output causes a failure (the unit tests only tests test vectors one way - I will fix that oversight 😁).
   - The code obviously needed to test greater than 64-byte results. I only had the spec test vectors to go by and the spec _still_ only has 32-byte tag test vectors.

**Comment by mheyman** [comment]
Found it! The issue was the length passed into in the final blake2 block. That was a _fun_ bug.

It is a breaking change unfortunately. Since I'm breaking things, I removed the $data field from the argon hash string for the additional data input. This must have been a pre-releasism I included back when I first wrote this code that got elided before the real reference code was released.

The code now builds the reference command line utility (slightly modified with some additional fields the original didn't allow setting). It then creates 5-600 "reference" vectors to check against (looking at corner cases) in a C# code generation build step. I'm pretty confident this code now behaves correctly at all hash lengths (the code does handle >2GB hashes on C#). I'm disappointed I didn't do this back at the start because I'm usually more aware of testing crypto code properly.

Currently I have a certificate issue pushing a nuget package - working on that...
